include/openPMD/Series.hpp

```diff
@@ -282,5 +348,139 @@ class Series : public Attributable
+        std::make_shared< auxiliary::Option< WriteIterations > >();
 }; // Series
+
+/**
+ * @brief Subclass of Iteration that knows its own index withing the containing
+ *        Series.
+ */
+class IndexedIteration : public Iteration
+{
+    friend class SeriesIterator;
+
+public:
+    using iterations_t = decltype( Series::iterations );
+    using index_t = iterations_t::key_type;
+    index_t const iterationIndex;
+
+private:
+    template< typename Iteration_t >
+    IndexedIteration( Iteration_t && it, index_t index )
+        : Iteration( std::forward< Iteration_t >( it ) )
+        , iterationIndex( index )
+    {
+    }
+};
+
+class SeriesIterator
+{
+    using iteration_index_t = IndexedIteration::index_t;
+
+    using maybe_series_t = auxiliary::Option< Series * >;
+
+    maybe_series_t m_series;
+    iteration_index_t m_currentIteration = 0;
+
+    // construct the end() iterator
+    SeriesIterator();
+
+public:
+    SeriesIterator( Series * );
+
+    SeriesIterator &
+    operator++();
+
+    IndexedIteration
+    operator*();
+
+    bool
+    operator==( SeriesIterator const & other ) const;
+
+    bool
+    operator!=( SeriesIterator const & other ) const;
+
+    static SeriesIterator
+    end();
+};
+
+/**
+ * @brief Reading side of the streaming API. Create instance via
+ *        Series::readIterations(). For use in a C++11-style foreach loop.
+ *        Designed to allow reading any kind of Series, streaming and non-
+ *        streaming alike. Calling Iteration::close() manually before opening
+ *        the next iteration is encouraged and will implicitly flush all
+ *        deferred IO actions. Otherwise, Iteration::close() will be implicitly
+ *        called upon SeriesIterator::operator++(), i.e. upon going to the next
+ *        iteration in the foreach loop.
+ *
+ *        Since this is designed for streaming mode, reopening an iteration is
+ *        not possible once it has been closed.
+ *
+ */
+class ReadIterations
+{
+    friend class Series;
+
+private:
+    using iterations_t = decltype( Series::iterations );
+    using iterator_t = SeriesIterator;
+
+    Series * m_series;
+
+    ReadIterations( Series * );
+
+public:
+    iterator_t
+    begin();
+
+    iterator_t
+    end();
+};
+
+/**
+ * @brief Writing side of the streaming API. Create instance via
+ *        Series::writeIterations().
+ *        For use via WriteIterations::operator[]().
+ *        Designed to allow reading any kind of Series, streaming and non-
+ *        streaming alike. Calling Iteration::close() manually before opening
+ *        the next iteration is encouraged and will implicitly flush all
+ *        deferred IO actions. Otherwise, Iteration::close() will be implicitly
+ *        called upon SeriesIterator::operator++(), i.e. upon going to the next
+ *        iteration in the foreach loop.
+ *
+ *        Since this is designed for streaming mode, reopening an iteration is
+ *        not possible once it has been closed.
+ *
```

Comment:
Curious formatting, you can do title + description like this:
```suggestion
/** Writing side of the streaming API.
 *
 * Create instance via Series::writeIterations().
 * For use via WriteIterations::operator[]().
 * Designed to allow reading any kind of Series, streaming and non-
 * streaming alike. Calling Iteration::close() manually before opening
 * the next iteration is encouraged and will implicitly flush all
 * deferred IO actions. Otherwise, Iteration::close() will be implicitly
 * called upon SeriesIterator::operator++(), i.e. upon going to the next
 * iteration in the foreach loop.
 *
 * Since this is designed for streaming mode, reopening an iteration is
 * not possible once it has been closed.
```